File: runner/runner/voice/match.py
"""Which open quest a spoken name means. Mirrors apps/web/src/voice/match.ts; both are locked to
packages/schema/fixtures/voice.json."""

from __future__ import annotations

import re
from collections.abc import Iterable
from dataclasses import dataclass
from typing import Literal

from runner.voice.grammar import normalize

STOPWORDS = frozenset(
    [
        "the",
        "a",
        "an",
        "my",
        "to",
        "of",
        "and",
        "for",
        "task",
        "quest",
        "el",
        "la",
        "los",
        "las",
        "lo",
        "un",
        "una",
        "mi",
        "mis",
        "de",
        "del",
        "al",
        "y",
        "tarea",
        "mision",
    ]
)
MIN_SCORE = 0.5


@dataclass(frozen=True)
class QuestMatch:
    kind: Literal["match", "ambiguous", "none"]
    ids: tuple[str, ...] = ()

    def as_dict(self) -> dict:
        if self.kind == "match":
            return {"kind": "match", "id": self.ids[0]}
        if self.kind == "ambiguous":
            return {"kind": "ambiguous", "ids": list(self.ids)}
        return {"kind": "none"}


def _words(s: str) -> list[str]:
    return [w for w in re.split(r"[^a-z0-9ñ]+", normalize(s)) if w and w not in STOPWORDS]


def _same(a: str, b: str) -> bool:
    """Same word, or one a prefix of the other (impuesto / impuestos) once both have 4+ letters."""
    return a == b or (min(len(a), len(b)) >= 4 and (a.startswith(b) or b.startswith(a)))


def match_quest(ref: str, quests: Iterable[tuple[str, str]]) -> QuestMatch:
    """quests: (id, title) pairs. Score = share of the spoken words found in the title; the best
    unique score >= 0.5 wins."""
    said = _words(ref)
    if not said:
        return QuestMatch("none")
    best = 0.0
    ids: list[str] = []
    for quest_id, title in quests:
        words = _words(title)
        hits = sum(1 for w in said if any(_same(w, t) for t in words))
        score = hits / len(said)
        if score < MIN_SCORE:
            continue
        if score > best:
            best, ids = score, [quest_id]
        elif score == best:
            ids.append(quest_id)
    if not ids:
        return QuestMatch("none")
    return QuestMatch("match" if len(ids) == 1 else "ambiguous", tuple(ids))

File: apps/web/src/voice/match.ts
// Which open quest a spoken name means ("complete laundry" -> "Do the laundry"). Pure code (P2);
// mirrors runner/runner/voice/match.py, both locked to packages/schema/fixtures/voice.json.

import { normalize } from "./grammar.ts";

export type QuestMatch =
  { kind: "match"; id: string } | { kind: "ambiguous"; ids: string[] } | { kind: "none" };

const STOPWORDS = new Set(
  (
    "the a an my to of and for task quest " +
    "el la los las lo un una mi mis de del al y tarea mision"
  ).split(" "),
);

function words(s: string): string[] {
  return normalize(s)
    .split(/[^a-z0-9ñ]+/)
    .filter((w) => w && !STOPWORDS.has(w));
}

/** Same word, or one a prefix of the other (impuesto / impuestos) once both have 4+ letters. */
function same(a: string, b: string): boolean {
  if (a === b) return true;
  return Math.min(a.length, b.length) >= 4 && (a.startsWith(b) || b.startsWith(a));
}

const MIN_SCORE = 0.5;

/** Score = share of the spoken words found in the title; the best unique score >= 0.5 wins. */
export function matchQuest(ref: string, quests: { id: string; title: string }[]): QuestMatch {
  const said = words(ref);
  if (said.length === 0) return { kind: "none" };
  let best = 0;
  let ids: string[] = [];
  for (const q of quests) {
    const title = words(q.title);
    const hits = said.filter((w) => title.some((t) => same(w, t))).length;
    const score = hits / said.length;
    if (score < MIN_SCORE) continue;
    if (score > best) {
      best = score;
      ids = [q.id];
    } else if (score === best) {
      ids.push(q.id);
    }
  }
  if (ids.length === 0) return { kind: "none" };
  return ids.length === 1 ? { kind: "match", id: ids[0] } : { kind: "ambiguous", ids };
}
